staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go

```diff
@@ -238,9 +239,9 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
 	}
 	if len(errs) > 1 {
 		for i := 1; i < len(errs); i++ {
-			// TODO: merge status errors; until then, just return the first one.
 			utilruntime.HandleError(errs[i])
 		}
+		return apierrors.NewInternalError(utilerrors.NewAggregate(errs))
```

Comment:
Thanks for the PR, but a 500 error is not correct here, and this masks / discards the type / status of the errors the webhooks returned.

For example, if webhooks returned [Invalid, Forbidden, Invalid, BadRequest], this would aggregate that into a 500 error.

I think we should only attempt to merge actual status errors returned from the webhooks (`err = rejectionErr.Status`), and can only accurately merge errors of the same type / status

Reply:
> I think we should only attempt to merge actual status errors returned from the webhooks (err = rejectionErr.Status), and can only accurately merge errors of the same type / status

It's unfortunate, but with data model, I don't see any alternative.

If there are multiple errors, I think it's important to prioritize any internal errors, because these errors may remain despite any changes the client makes to the request.

So, something like:

If there are two or more errors, then group them, merging the details, by Reason (e.g. InternalError, Forbidden, Invalid, BadRequest, etc.), and order the groups by Reason, starting with InternalError. Return the first group.

Reply:
> hmm... I'd actually prioritize returning errors that are actionable for the user (BadRequest, Invalid, Forbidden, etc) if we can't merge them all

Okay. I can see some reasons for this: internal errors that are transient, or being aware of some required changes, even if some webhooks failed to run.

Reply:
@dlipovetsky I'll continue working on this PR.

Reply:
I submitted a version. How do you think about this implementation?